SSR server side login for firebase client SDK

mentAl-maZe opened the pull request:
New configuration for `firebase.services.auth.initialize.ssr` property:

```ts
firebase.services.auth.initialize.ssr = boolean | {
  credential?: string
  serverLogin?: boolean
}
```
- `credential`:  
  takes either a nuxt.resolver path (e.g. '~/<path>')  
  or a googleServiceAccount object

  If this is omitted the module falls back on using the `admin.credential.applicationDefault()`
- `serverLogin`:  
  Server login is disabled by default to avoid API issues (firebase auth rate limiting by IP)

If you have intetion to merge this please let me know so I'll update the documentation with the description for the new configuration and a comment about the rate limiting caveat.

lupas commented on lib/helpers/createServerMiddleware.js:
```diff
@@ -0,0 +1,47 @@
+      // separate properties preserve established API
+      const { customClaims, ...authUser } = userRecord
+
+      // @TODO: deprecate either response or request object augmentation on next major version bump
```
I reckon you prefer using `req` rather than `res` to pass on data.

I was considering `res.locals` because according to [here](https://stackoverflow.com/questions/18875292/passing-variables-to-the-next-middleware-using-next-in-express-js) this is the recommended way to do it.

Others prefer `req`. 

I assume `res.locals` would be the most official way to do it, but `req` also works I guess. Then again I'm not a NodeJs specialst.

Any arguments for either or?

mentAl-maZe replied:
I consider authentication & authorization to be request augmentations for that reason i prefer using `req` to set resolved users on.

The way I see it, it belongs on the `req` because logically it is information concerning the issued request, not the response, i.e. user X (`verifiedFireAuthUser`) has issued the request (`req`) and should therefore be associated with the request object.

However it's your module (thanks for the great work, btw :sweat_smile: ) so the decision is yours.

lupas replied:
@mentAl-maZe 
Yes that kind of makes sense to me, too.

At the same time [here](http://expressjs.com/en/api.html#res.locals) the Express team uses `res.locals`, and not `req.locals` or `req` for passing exactly these kind of authentication data, no?

> _exposing request-level information such as the request path name, authenticated user, user settings, and so on._"

Hmm... I'm happy with either or. Small detail, we can also decide to just do both like it is now ;-)

> However it's your module (thanks for the great work, btw 😅 ) so the decision is yours.

Bah, it's open source, I do listen to people who know better :P

mentAl-maZe replied:
> At the same time here the Express team uses res.locals, and not req.locals or req for passing exactly these kind of authentication data, no?

I guess I'll have to reconsider than.
I'm all for community standards and if this is how the core team of the underlying technology recommends doing it I think we should follow suit.

lupas commented:
This looks pretty great and solves 2 current issues at once - great job! 🎉 

I will go and test it thoroughly now and if it all works as expected I'm more than happy to do the merge.

### Open tasks:

- [ ] Test (@lupas)
- [ ] Update documentation (@mentAl-maZe) - Thanks!
- [ ] Decide on `req`, `res` or `res.locals` (@mentAl-maZe your opinion?)
- [ ] If we decide on making it a major (breaking) update because of the former point, I would also like to include the fix for https://github.com/nuxt-community/firebase-module/issues/118 in the same release so we have it all in one update (@lupas)
- [ ] If so: Write short upgrade guide (@lupas)

mentAl-maZe commented:
Great :grin: 
I'll update the documentation later today if i manage, otherwise i'll do it over the weekend

lupas commented:
@mentAl-maZe 

>  ERROR  logger is not a function                                                                           17:06:52
>  at /Users/{foo}/lib/helpers/createServerMiddleware.js:42:7

mentAl-maZe commented:
Should be fixed in [latest commit](https://github.com/nuxt-community/firebase-module/pull/121/commits/82e1a82763b8cb63fc32b17c0fb89dae39ab2c02)

mentAl-maZe commented:
BTW: If you intend on releasing a major version, i'd suggest the following API changes

- `res.locals.user` (as recommended by express docs) holds complete `admin.auth.UserRecord` no need for an additional claims property
- additional `services.auth.initialize.ssr.ingorePaths: (string | RegEx)[]` property to avoid triggering serverMiddleware (useful for use with APIs #120)

let me know what you think

lupas commented:
> @mentAl-maZe 
> Should be fixed in latest commit

Silly me pulled too early. Great!

## One more thing:

- [ ] `ssr: true` throw errors due to missing credentials on non-firebase infrastructure

>  ERROR  Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: invalid_grant (Bad Request)". There are two likely causes: (1) your server time is not properly synced or (2) your certificate key file has been revoked. To solve (1), re-sync the time on your server. To solve (2), make sure the key ID for your key file is still present at https://console.firebase.google.com/iam-admin/serviceaccounts/project. If not, generate a new key file at https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk.

One of the nice things with the JWTDecode solution was that we could set `ssr: true` and got a verified result **without the need to set admin credentials**. It would be great if that would still be possible.

- If `ssr: true`: Just user verification **without** the need to set credentials
- If `ssr: { credentials: true }` use credentials from applicationDefault().
- If `ssr: { credentials: 'someString' }` use credentials from string.

For that the server-middleware would have to look somewhat like this:

**No-Credentials:**
```js
admin.initializeApp(configWithoutCredentials)
await admin.auth().verifyIdToken(idToken, false)
const userRecord = JWTDecode(idToken)
```
**Credentials:**
```js
admin.initializeApp(configWithCredentials)
await admin.auth().verifyIdToken(idToken, true)
const userRecord = await admin.auth().getUser(decodedToken.uid)
```

Of course the code could be written more nicely. Do you think that would make sense to implement? Or will it overcomplicate things?

lupas commented:
> BTW: If you intend on releasing a major version, i'd suggest the following API changes
> 
> * `res.locals.user` (as recommended by express docs) holds complete `admin.auth.UserRecord` no need for an additional claims property
> * additional `services.auth.initialize.ssr.ingorePaths: (string | RegEx)[]` property to avoid triggering serverMiddleware (useful for use with APIs #120) and it could solve the 'too many auth requests' issue
> 
> let me know what you think

Makes both sense to me, good stuff! :)

mentAl-maZe commented:
> ### One more thing:
> `ssr: true` throw errors due to missing credentials on non-firebase infrastructure

I didn't realize the missing credential was what made it work :confused: 

> No-Credentials:
> ```js
> admin.initializeApp(configWithoutCredentials)
> await admin.auth().verifyIdToken(idToken, false)
> const userRecord = JWTDecode(idToken)
> ```

Doesn't `admin.auth().verifyIdToken(idToken, true)` already return the decoded token, regardless of credential setting?

In that case we could avoid the additional dependecy on `JWTDecode`

I'll make some tests.

lupas commented:
> Doesn't `admin.auth().verifyIdToken(idToken, true)` already return the decoded token, regardless of credential setting?
> 
> In that case we could avoid the additional dependecy on `JWTDecode`
> 
> I'll make some tests.

Apparently, if set to true, the admin-sdk needs to have credentials set and `verifyIdToken()` fails. If it is set to false, it seems to work without credentials.

At least that's what my tests showed. `getUser()` always needs credentials.

mentAl-maZe commented:
Ok, so the revocation check fails but it should still return the decoded token right?

lupas commented:
> Ok, so the revocation check fails but if we don't check revocation it should still return the decoded token right?

Well that is how it currently works. We might need to mention in the docs that if no credentials are set, that the user is then not checked for revocation.

lupas commented on lib/helpers/createServerMiddleware.js:
```diff
@@ -1,20 +1,77 @@
 const admin = require('firebase-admin')
 const logger = require('../logger')
 
+const simulateUserRecord = ({
```
Won't have much time tonight to look at it in more detail, but is there a way to avoid this?

This will make the maintenance of this module very cumbersome. When Firebase changes just a small thing here, we will always have to rush and change the same in the module, too.

Furthermore it makes it a pain for anyone who wants to contribute. The more code, the less people will wanna look at it.

If possible I would like to avoid something like this...

What's your opinion? It might even be a good reason to just drop the entire "no-credential" idea.

hmm..

mentAl-maZe replied:
:thinking: 
I see your point...
I just did this to provide a stable API to the user, but your quite right.

What about we just guarantee `{ uid, email, emailVerified, displayName }` and put all additional claims into `{ /* ... */, customClaims }`?

lupas replied:
Hmm yeah... 🤔

I see that it would make sense to provide a stable API and return the same object, but on the other hand this module should also be stable in the future without us having to intervene after every Firebase upgrade.

> What about we just guarantee { uid, email, emailVerified, displayName }
Maybe that's a good trade-off. And if someone needs all data there is always the option to provide credentials. 
> and put all additional claims into { /* ... */, customClaims }?
You mean you want to mix it with custom claims?

mentAl-maZe replied:
> You mean you want to mix it with custom claims?

`admin.auth.DecodedIdToken` is just the decoded JWT Token with an added `uid` property, custom claims live on the root of this object as they do on the object returned by `firebase.auth().currentUser.getIdTokenResult()` in that sense it might even be desirable as it connects better to the `onAuthStateChangedMutation/Action`

lupas replied:
> > You mean you want to mix it with custom claims?
> 
> `admin.auth.DecodedIdToken` is just the decoded JWT Token with an added `uid` property, custom claims live on the root of this object as they do on the object returned by `firebase.auth().currentUser.getIdTokenResult()` in that sense it might even be desirable as it connects better to the `onAuthStateChangedMutation/Action`

Oh, right, I forgot. Alright, then let's do it that way :) 👍

mentAl-maZe replied:
I have opted to create a custom user object with the following specification:

```d.ts
export type FireAuthServerUser = Omit<auth.UserRecord, 'disabled' | 'metadata' | 'providerData'>
  & Partial<Pick<auth.UserRecord, 'disabled' | 'metadata' | 'providerData'>>
  & { allClaims: auth.DecodedIdToken }
```
Essentially we provide a "light" version of the `auth.UserRecord` without `credential` and augment to full if `credential` is set.
The `admin.auth.DecodedIdToken` is set on a separate property (`allClaims`) so it remains consistent across both implementations.

Let me know what you think.

mentAl-maZe commented:
No worries, I need this for my project anyways :sweat_smile: 

Deal, will implement as recommended and update the documentation.

mentAl-maZe commented:
On second thought, most users currently will expect the `authUser` and `claims` objects to be defined in their `onSuccessMutation`/`onSuccessAction`.

It is a breaking API change in any case, either new property name or new variable state.

The least disruptive way would be not to rename the `onSuccessMutation`/`onSuccessAction` properties and add new `onLogoutMutation`/`onLogoutAction` properties.

How do you want me to proceed?

mentAl-maZe commented:
### Updated Tasks:

- [ ] Tests
- [ ] Update documentation
- [ ] Write short upgrade guide
- [x] `ssr: true` throw errors due to missing credentials on non-firebase infrastructure
  > implemented `admin.auth.UserRecord` simulation
- [x] Decide on `req`, `res` or `res.locals`
  > `res.locals.user = admin.auth.UserRecord`
- [x] fix for #118 
  > call `onSuccessMutation`/`onSuccessAction` with `{ authUser: null, claims: null }`

lupas commented:
> How do you want me to proceed?

I see your point and it makes perfect sense..

BUT at this point I rather have **less complexity** in the module than a release being less disruptive, the module is still young and I have confidence in our users that they prefer a small one-time change in their code to the module becoming more and more complex.

Go for `onAuthStateChangedAction` and `onAuthStateChangedMutation`, it will make the option so much more self-explanatory.

Oh and by the way:
Do you think we even need `_handleAuthError()` and the respective mutation/action? If I look at the code now I kinda feel like they are a bit redundant. What do you think?

mentAl-maZe commented:
> Do you think we even need _handleAuthError() and the respective mutation/action? If I look at the code now I kinda feel like they are a bit redundant. What do you think?

In the end a user could implement error handling himself in his custom `onAuthStateChangedAction`/`onAuthStateChangedMutation`

Although `authUser.getIdTokenResult(true)` might throw without missing connectivity

lupas commented:
> Although `authUser.getIdTokenResult()` might throw with missing connectivity

But we are currently not catching any errors from getIdTokenResult() 👀 . I recollect reading somewhere that `getIdTokenResult()` is only carried out on the  client anyway, not sure if there can be connectivity issues...

mentAl-maZe commented:
> But we are currently not catching any errors from getIdTokenResult() eyes . I recollect reading somewhere that getIdTokenResult() is only carried out on the client anyway, not sure if there can be connectivity issues...

I guess you're right even in a PWA with cached data the handler should never receive a stale `firebase.User` object as it is passt directly from `onAuthStateChanged()`, so error handling could be done entirely by the user himself in the `onAuthStateChangedAction` (`onAuthStateChangedMutation` will never throw an error unless the user messes up the mutation and then an error is justified)

OK, I'll remove the `onAuthErrorMutation/Action`

mentAl-maZe commented:
I fell into the rabbit hole and finalized the release :blush: 

## Feature List:

- removed unnecessary `onAuthErrorMutation/Action`
- renamed `onAuthSuccessMutation/Action` to more descriptive `onAuthStateChangedMutation/Action`
- `onAuthStateChangedMutation/Action` now is called for `authUser = null` as well
- added `ssr.credential` configuration to authorize firebase-admin
- implemented `ssr.serverLogin` for hydration operations
- service worker no longer augments requests to static resources and HMR calls.
- Additionally ignorePaths can be set as `'/string'` (`pathname.startsWith('/string')`) or `/RegExp/` (`/RegExp/.test(pathname)`) 

## Updated Tasks @lupas 

- [ ] Confirm implementation (mainly `FireAuthServerUser`)
- [ ] Check documentation (`/guide/options#auth` and `/tutorials/ssr`)
- [ ] Tests (@mentAl-maZe test in production)

lupas commented on docs/guide/options/README.md:
```diff
@@ -204,23 +209,85 @@ export const mutations = {
   onSuccessMutation: (state, { authUser, claims }) => {
```
rename to `ON_AUTH_STATE_CHANGED_MUTATION`

mentAl-maZe replied:
Done (to be committed)

lupas commented on docs/guide/options/README.md:
```diff
@@ -204,23 +209,85 @@ export const mutations = {
+Server is stateless and has to sign in on every request.  
+Firebase client SDK login attempts are rate limited by IP, multiple server side calls will cause an 'auth/too-many-requests' error.
+
+If you have an API which is served over nuxt ssr, add it's prefix to `auth.ssr.ignorePaths`
```
is this sentence complete?

mentAl-maZe replied:
I improved the description and added an example (to be committed)

mentAl-maZe replied:
I'm currently waiting for response from firebase community and have not yet implemented load tests.
Once I have a clearer understanding of the issue I'll update the warning accordingly

lupas commented on docs/tutorials/ssr/README.md:
```diff
@@ -1,84 +1,135 @@
-},
+
+  // use custom claims (see https://firebase.google.com/docs/auth/admin/custom-claims)
+  if (claims.admin) {
```
Two points:

- Custom claims don't need to be used necessarily in this action, it's just an example what one could do. The action can also be used to do other asynchronous calls not related to custom claims at all.

- The "admin" custom claim is not a given property and may not even exist / be used in many projects.

Due to this I would prefer to **not** use it as an example everywhere in the docs and just write something like `// do other async calls`, just for not confusing people with this.

(also in the other examples further up)

mentAl-maZe replied:
I have renamed the requested property to `claims.custom_claim` and added a reference to the [corresponding firebase guide](https://firebase.google.com/docs/auth/admin/custom-claims).

I feel the tutorial should provide reference material where possible.

Let me know what you think

lupas replied:
All good - thanks. 👍

lupas commented on docs/tutorials/ssr/README.md:
```diff
@@ -1,84 +1,135 @@
-      authUser: ssrVerifiedAuthUser,
-      claims: ssrVerifiedAuthUserClaims
+async nuxtServerInit({ dispatch, commit }, { res }) {
+  if (res && res.locals && res.locals.user) {
```
can `res` even be be null? `if (res.locals && res.locals.user)` should do, no?

lupas replied:
Forget my comment. In `nuxt generate` mode it does indeed not exist.

lupas commented on docs/tutorials/ssr/README.md:
````diff
@@ -1,84 +1,135 @@
 ```bash
-yarn add firebase-admin # OR npm i firebase-admin
-yarn add jwt-decode # OR npm i jwt-decode
+yarn add firebase-admin @nuxtjs/pwa # OR npm i firebase-admin @nuxtjs/pwa
````
can we add this change to do **upgrade-guide**, too?

mentAl-maZe replied:
Done (please check)

lupas commented on docs/tutorials/ssr/README.md:
````diff
@@ -1,84 +1,135 @@
 ```js
-handleSuccessfulAuthentication({ commit }, { authUser }) {
-  // Install servicerWorker if supported on sign-in/sign-up page.
-  if (process.browser && 'serviceWorker' in navigator) {
````
can we add this to the **upgrade-guide**, too?

- "Service worker registration new not in the handleSuccessfulAuthentication action, but through @nuxtjs/pwa"

mentAl-maZe replied:
Done (please check)

lupas commented on docs/guide/upgrade/README.md:
```diff
@@ -1,4 +1,111 @@
-# Upgrade from v3 to v4
+# Migration guides
+
+## Upgrade from v4 to v5
```
Add:

- If using the `ssr: true` mode, rename `ctx.res.verifiedFireAuthUser` to `ctx.res.locals.user` in your **nuxtServerInit**.

Maybe add a comment in general saying:

- We have completely changed how SSR Authentication works. If you have set up SSR like previously mentioned [here](http://localhost:8080/tutorials/ssr), follow the Tutorial again to properly set it up with v5.

mentAl-maZe replied:
Done (please check)

lupas commented on lib/templates/firebase-auth-sw.js:
```diff
@@ -40,56 +41,56 @@ const getIdToken = () => {
+      mode: 'same-origin'
+    }));
+  } catch (e) {
+    // This will fail for CORS requests. We just continue with the
```
I'm getting:

> firebase-auth-sw.js:90 Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'WorkerGlobalScope': 'only-if-cached' can be set only with 'same-origin' mode
    at firebase-auth-sw.js:90

At this line:

>   event.respondWith(getIdToken().then(idToken => idToken ? requestProcessor(idToken, event.request) : fetch(event.request)));

Expected on localhost?

Only on initial visit after clearing cache, afterwards everything else works as expected.

mentAl-maZe replied:
Could you set a breakpoint on the line and check the request payload that produces the error?

lupas replied:
unable to reproduce myself now :)

lupas commented on lib/plugins/initAuth.js:
```diff
@@ -1,51 +1,27 @@
+      const claims = authUser ? (await authUser.getIdTokenResult()).claims : null
+
+      if (onAuthStateChangedMutation) {
+        store.commit(onStateChangedMutation, { authUser, claims })
```
!!!
Should be on**Auth**StateChangedMutation

mentAl-maZe replied:
Done (to be committed)

lupas commented:
So I just tested your changes and it works well....
Except for one major issue.

I tested the following:

- SSR Auth without credentials
  - [X] successfully received the simplified user record 
- SSR Auth with credentials
  - [X] successfully received the full user record (with `serverLogin: false`)
  - [X] successfully received the full user record (with `serverLogin: true`)
  - [X] successfully fetched secured data with `serverLogin: true`
 
 **MAJOR PROBLEM:**
  - Was able to fetch the same secured data with another incognito window & another browser that was previously not logged in

This is kind of what we suspected with the serverLogin part, see [here](https://github.com/nuxt-community/firebase-module/issues/117#issuecomment-597648407). It seems that the server shares its auth state among requests. That's what we initially wanted to avoid - and is quite critical.

Do you see a way around this?

Can we log out after each request? Probably critical too...

Hmm...

But all the other features and changes are great and work as expected! :)

mentAl-maZe commented:
:scream: 
I didn't consider this, good thing this came up before i deployed my project... (community testing ftw :sweat_smile: )

In theory calling `await firebase.auth().setPersistence(firebase.auth.Persistence.NONE)` should remedy the problem, but I'm currently testing an implementation where I only change auth state if a user is not already signed in, might help reduce authorization attempts.

```js
if (app.$fireAuth.currentUser) {
  if (app.$fireAuth.currentUser === res.locals.user) {
    return
  }
  await app.$fireAuth.signOut()
}

// create custom token and sign in
```

lupas commented:
> 😱
> I didn't consider this, good thing this came up before i deployed my project... (community testing ftw 😅 )
> ```

😆 yeah that's the main issue with using the client SDK on the server-side. Yet have to find a way around it.

> In theory calling await firebase.auth().setPersistence(firebase.auth.Persistence.NONE) should remedy the problem

I don't think it would. Remember, there is only ONE server, so it does not matter what the user does/suers do. When the server is not shut down, the authState should also persist if persistence is set to NONE.

I'm not sure if all of these approaches will work, because even if you sign out after the request, there might still be the chance of another request snitching these auth credentials if the timing is right, no? Not sure how secure that would be... :o

mentAl-maZe commented:
> (and you DON'T want to take chances when it comes to authentication)

You're absolutely right! :+1: 

> I don't think it would. Remember, there is only ONE server, so it does not matter what the user does/suers do. When the server is not shut down, the authState should also persist if persistence is set to NONE.

You're right an unauthorized user might hitch a ride on the session of an authorized user if he accesses the server at the same time... :thinking: 

> I don't think it would. Remember, there is only ONE server, so it does not matter what the user does/suers do. When the server is not shut down, the authState should also persist if persistence is set to NONE.
I'm not sure if all of these approaches will work, because even if you sign out after the request, there might still be the chance of another request snitching these auth credentials if the timing is right, no? Not sure how secure that would be... :o

Every SSR request spins up a vue instance so a plugin should be executed for every request (I'm not sure about nuxt's caching mechanisms), so this should work.

Although we could still run into concurrency issues (two users access the server at the same time) as the firebase instance is initialized outside of the server hydration flow.

**Possible solution:**
Import firebase dynamically in the plugin so every request gets its own firebase instance.
Will investigate...

lupas commented:
> > (and you DON'T want to take chances when it comes to authentication)
> 
> You're absolutely right! 👍
> 
> > I don't think it would. Remember, there is only ONE server, so it does not matter what the user does/suers do. When the server is not shut down, the authState should also persist if persistence is set to NONE.
> 
> You're right an unauthorized user might hitch a ride on the session of an authorized user if he accesses the server at the same time... 🤔
> 
> > I don't think it would. Remember, there is only ONE server, so it does not matter what the user does/suers do. When the server is not shut down, the authState should also persist if persistence is set to NONE.
> > I'm not sure if all of these approaches will work, because even if you sign out after the request, there might still be the chance of another request snitching these auth credentials if the timing is right, no? Not sure how secure that would be... :o
> 
> Every SSR request spins up a vue instance so a plugin should be executed for every request (I'm not sure about nuxt's caching mechanisms), so this should work.
> 
> Although we could still run into concurrency issues (two users access the server at the same time) as firebase is imported statically.
> 
> **Possible solution:**
> Import firebase dynamically in the plugin so every request gets its own firebase instance.
> Will investigate...

Or maybe there's a workaround using session cookies?
https://firebase.google.com/docs/auth/admin/manage-cookies

Maybe we can also implement a check in the serverMiddleware that blocks all requests to the server as 401 if the uid from the req.headers is not the same as the `auth.currentUser's` on the server..

mentAl-maZe commented:
> Or maybe there's a workaround using session cookies?

Session cookies are just another way of getting the id token to the server and i much prefer your service worker implementation for that.

The problem of a stateful Firebase client SDK remains...

> Maybe we can also implement a check in the serverMiddleware that blocks all requests to the server as 401 if the uid from the req.headers is not the same as the auth.currentUser's on the server..

That would block every other user for the duration of a request. (Pagecrashes aplenty)

## Possible solution 1
Import firebase dynamically in the plugin so every request gets its own firebase instance.

### Issue
Pages with heavy traffic will authenticate every page load (once the user is on the page, navigation is handled by the client).
After a certain amount of authorized page loads the server side authentication seizes to work because of `auth/too-many-auth-requests` error.

### Feasibility

This is closer to a solution.

But for applications with many authenticated users (making use of some sort of `remember me` functionality) a scenario where 100 authenticated page loads are triggered in a short amount of time because the users open a previously closed PWA/Browser tab, becomes very likely.

## Possible solution 2

Create a seperate app on the server side for every new UID.
In `plugin/main.js`
```js
export default async ({ res }, inject) => {
  let resolvedApp
  if (process.server && res && res.locals && res.locals.user) {
    resolvedApp = firebase.apps.find(a => a.name === res.locals.user.uid) || firebase.initializeApp(firebaseConfig, res.locals.user.uid)
  } else {
    resolvedApp = firebase.apps.find(a => a.name === '[DEFAULT]') || firebase.initializeApp(firebaseConfig)
 }

  <% if (options.services.auth) { %>
    <%= writeImportStatement('auth') %>

    const fireAuth = resolvedApp.auth()
    const fireAuthObj = firebase.auth
    inject('fireAuth', fireAuth)
    inject('fireAuthObj', fireAuthObj)
  <% } %>

  // ...
  }
})

### Issue
Possible memory leak as a buttload (pardon my french) of apps will be created with pages with heavy traffic and they are never deleted (I think firebase keeps an in memory cache of initialized apps).
Also I'm not sure this would resolve the rate limiting issue.

### Feasibility

This solves the concurrency problem, but it introduces a possible memory issue.

However if this also solves the rate limiting issue I think it would be worth to implement some sort of cleanup procedure in `this.nuxt.hook('generate:done')`

lupas commented:
Well it seems like we will have to make some sort of compromise anyway. Solution 2 sounds more like the official solution to me, if we find a clever way to clean up the apps regularly (i.e. delete app after request).

Doesn't option 1 come with a similar hit on memory?

mentAl-maZe commented:
:thinking: In theory yes, but without the need for cleanup afterwards :sweat_smile:
but this is all an assumption as of now, I'll get back to you as soon as I have implemented load tests on my project.

PS: I just rechecked the docs and I didn't find any mention about rate limiting of **successful** auth attempts (got in touch with firebase slack communtity to clarify)

mentAl-maZe commented:
> :thinking: In theory yes, but without the need for cleanup afterwards :sweat_smile: 
> but this is all an assumption as of now, I'll get back to you as soon as I have implemented load tests on my project.

I did some digging and found that **Solution 1** indeed creates a memory leak with no clear way to plug it... consequently **Solution 1** has been rejected.

**Solution 2** has been implemented with an additional `auth.ssr.sessionLifetime` setting to configure how long the server should keep an instantiated app per user.

Check it out and let me know what you think.

mentAl-maZe commented:
# Finalization

- [ ] Wait for response from firebase community and update documentation accordingly
- [ ] Perform load tests to certify feasibility for larger projects

lupas commented:
> > 🤔 In theory yes, but without the need for cleanup afterwards 😅
> > but this is all an assumption as of now, I'll get back to you as soon as I have implemented load tests on my project.
> 
> I did some digging and found that **Solution 1** indeed creates a memory leak with no clear way to plug it... consequently **Solution 1** has been rejected.
> 
> **Solution 2** has been implemented with an additional `auth.ssr.sessionLifetime` setting to configure how long the server should keep an instantiated app per user.
> 
> Check it out and let me know what you think.

Sounds great! I'm having a look at it now and let you know.

We might want to consider setting the ServerLogin feature as still "Experimental" until we thoroughly collected the evidence that it works properly in big projects.

lupas commented:
@mentAl-maZe 

I think you made it! 🎉 Congrats! This seems to work quite well!
I agree, we should run some more (load) tests and if all works, this seems to be ready to ship.

As mentioned earlier, maybe it's better if we still set the experimental flag for serverLogin so people are inclined to test it properly themselves before using in production. At least for a while.

Let me know when your finalization part is done, then we can merge this PR.

BTW:
I'm thinking of also implementing https://github.com/nuxt-community/firebase-module/issues/122 while we're at it.

lupas commented:
I don't know how exactly, but I just got the following error, I think after reloading an idle page after some minutes:

```js
 FATAL  Firebase: Firebase App named 'fooofooofoouserId' already deleted (app/app-deleted).        15:26:40

  at FirebaseAppImpl.checkDestroyed_ (node_modules/@firebase/app/src/firebaseApp.ts:173:27)
  at FirebaseAppImpl.get [as options] (node_modules/@firebase/app/src/firebaseApp.ts:85:10)
  at /Users/fooooooo/@nuxtjs-firebase/lib/module.js:76:48
  at node_modules/hable/lib/hookable.js:50:45
  at node_modules/hable/lib/utils.js:15:67
  at Nuxt.callHook (node_modules/hable/lib/hookable.js:50:7)
  at nuxtMiddleware (node_modules/@nuxt/server/dist/server.js:209:5)


   ╭────────────────────────────────────────────────────────────────────────────────────────────────╮
   │                                                                                                │
   │   ✖ Nuxt Fatal Error                                                                           │
   │                                                                                                │
   │   FirebaseError: Firebase: Firebase App named 'fooofooofoouserId' already deleted   │
   │   (app/app-deleted).                                                                           │
   │                                                                                                │
   ╰────────────────────────────────────────────────────────────────────────────────────────────────╯

error Command failed with exit code 1.
```

lupas commented on docs/guide/options/README.md:
```diff
@@ -158,69 +158,168 @@ auth: true
+
+::: warning Do not use firebase client SDK to perform API operations.  
+
+@TODO: Warn about special considerations when using client SDK on the server
```
don't forget this

mentAl-maZe commented:
Thanks :grin: 

Just one more thing
I'm currently testing with `firebase emulators:start` as this is the actual use case for my project (generated hosting with redirect to firebase functions for SSR)

With this setup the service worker is unable to resolve SSR requests with error
[nuxt-fire-sw-error.log](https://github.com/nuxt-community/firebase-module/files/4334434/nuxt-fire-sw-error.log)

mentAl-maZe commented:
I have updated the settings a bit to bundle experimental functionality, improved the service worker and implemented the persistence option as per issue #122 

I have requested the exact numbers of the QPS limits for client SDK authentication, once i hear back form them i'll try to improve the SSR serve login implementation or at least update the docs accordingly.

I still haven't gotten it to work with firebase functions SSR, I'll look into it tomorrow.
I think we will have to return to your `this.addTemplate` implementation for the server middleware and find another way around issue #87 

However all of this said, I think the current version is stable enough to be released, please check yourself and merge if everything is to your satisfaction.

I'll open a new pull request if my work with this module produces further improvements.

lupas commented:
@mentAl-maZe 

I agree, let's merge and release this for now. Thanks so much for your effort, this is awesome!
Also thanks for implementing https://github.com/nuxt-community/firebase-module/issues/122 !

I'll create issue(s) for the open points, would of course be great if we could find a solution for that. I would prefer **not** going back to the `this.addTemplate` solution, if there is a way around it.

Thanks a lot :) v5 will be released, soon.

mentAl-maZe commented:
You're welcome, always happy to help :smiley: 

I'll comment in the pertinent issues once i have made any progress.
